src/obfl-specification.html

```diff
@@ -1628,6 +1652,9 @@ <h4 id="L9241" class="include">toc-sequence</h4>
           <dd>Specifies a translation policy.</dd>
           <dd>See the <a href="#L2462">attribute description</a> for more
             information.</dd>
+        <dt>apply-toc-entry-content-use-when [optional]</dt>
+          <dd>Whether or not to use the toc-entry-content elements in the corresponding table-of-contents that have a use-when attribute.</dd>
+          <dd>One of 'false' (default) or 'true'.</dd>
```

Comment:
Two remarks:

- I vaguely remember us talking about using the `range` attribute to control whether the document range toc should include the continuation entries or not. However in principle a document range `toc-sequence` can reference another `table-of-contents` than the volume range `toc-sequence`. The resulting OBFL will be less beautiful, but it's not really a problem. Not sure if I realized this before when I brought up the idea.
- What's problematic with this solution is that an option to disable all the `use-when` is weird. Better would be an option to disable the `$resumed` variable (set it to a constant false). The name is also really wordy :-) but OK let's forget that for a moment.
- I still wonder whether it would make more sense to add new `range` values instead of a dedicated attribute, because basically it is the range that we are controlling. There would be a total of 4 `range` values:
  - document
  - volume
  - document-incl-resumed
  - volume-incl-resumed

Reply:
I like the idea of 4 range variables.
I would also suggest to use a `toc-entry-content-resumed` element instead of the `use-when="$resumed"` construction. I have a feeling that this 'use-when` is not coming naturally especially because the only option that we will support is `$resumed`. 
Also, note that OBFL has already the concepts `any-interrupted` and `any-resumed`. These concepts make no use of a `use-when`. I mean, it is `any-resumed` and not `any use-when="$resumed"`.

Reply:
Hum are we going back to that discussion? I though it was sorted.

I understand the concern that the `use-when` it is a bit artificial because there are no other use cases. But I have nothing against the principle. I would also be fine with a dedicated element, I don't have a strong opinion about it (although if we go for the dedicated element I would go for `toc-entry-content-when-resumed` or `toc-entry-content-on-resumed`).

The comparison with `any-interrupted` etc. is not totally appropriate I think. These elements are not part of a switch. Together they form a single "volume-transition", evaluated once (not twice) per volume. But I understand where you're coming from, it's indeed related in a way. It depends on how you explain how these features work exactly.

Reply:
The point is that a `use-when` only determines _when_ a `toc-entry-content` may be used, but it is impossible to add any specific functionality. This is not enough. It is too general.

For instance, if you look at Example 1 in https://github.com/brailleapps/dotify.formatter.impl/issues/106#issuecomment-543045156, case 4, you see a scenario where the `toc-text-resumed` of a parent is taken. This cannot be expressed with merely a `use-when` construction.

@fredrikschill I like your idea of keeping the OBFL specification as general as possible. However in this case we need a more specific construction. I will adapt @bertfrees's suggestion `toc-entry-content-on-resumed`.

Reply:
> For instance, if you look at Example 1 in [brailleapps/dotify.formatter.impl#106 (comment)](https://github.com/brailleapps/dotify.formatter.impl/issues/106#issuecomment-543045156), case 4, you see a scenario where the `toc-text-resumed` of a parent is taken. This cannot be expressed with merely a `use-when` construction.

OK true, but I'm not really convinced that it should behave like in that example. We could also require that in those cases you explicitly repeat the `toc-entry-content use-when="..."` in the descendant `toc-entry`s that are supposed to inherit. Sometimes it's good to leave out special behavior, it can make the spec easier to understand.

Anyway, like I said I don't have a strong opinion about which road to take. @kalaspuffar @PaulRambags @fredrikschill I'll let you guys come to an agreement.

Reply:
Hi @PaulRambags, we cannot see why case 4 in https://github.com/brailleapps/dotify.formatter.impl/issues/106#issuecomment-543045156 would not work for you when we use the "use-when" attribute. Can you please show this with an updated example for us all? Thanks!

Reply:
@fredrikschill It's about the inheritance. You can not mix in inheritance when you have a switch for every entry.

Reply:
@fredrikschill @kalaspuffar  Here is an updated example, in which I illustrate that a `use-when` construction with a `$resumed` variable is not working.

**Example**
Consider this picture which shows the sections of a book, with a volume break.

Volume 1
1
1.1
1.1.1
1.1.2
1.1.3
1.2
1.2.1
1.2.2
2
2.1
2.1.1
2.1.2

Volume 2
2.2 (which does not start at the beginning of the volume,
    so there is some resumed content of the previous volume,
    in other words, `$resumed=true`)
2.2.1
2.2.2
2.3

Now, the situation is such that only the top level `toc-entry`s have a resumed text.
That is, only section 1 and section 2 have a resumed text, which are "Continuation of Chapter 1" and "Continuation of Chapter 2", respectively. Sections 1.1, 1.1.1, 1.1.2, etc. have no resumed text. See below for a precise detailed description.

At the start of Volume 2, the text "Continuation of Chapter 2" must be inserted.

With a use-when construction and a `$resumed` variable, it is possible to describe that a resumed text must be inserted before section 2.2, because the content does not start at the beginning of the volume. In other words, there is still some resumed content of the previous volume. Therefore `$resumed=true`.
However it is not possible to indicate that NOT the resumed text of the previous entry (which is 2.1.2 and it has no resumed text) must be inserted, instead the `toc-entry` hierarchy of the previous entry must be descended until a resumed text is found.
In this case, 2.1.2 has no resumed text. Its parent in the `toc-entry` hierarchy is 2.1 which also has no resumed text. The parent of 2.1 in the toc-entry hierarchy is 2 which has a resumed text, viz. "Continuation of Chapter 2". 
Bert calls this _inheritance_. In his last remark he says that the resumed text is _inherited_ from a parent.

Detailed description:
The above is a simplified picture of this toc-sequence:

```
<toc-sequence range="volume-incl-resumed">
  <toc-entry ref="1">
    <toc-entry-content>Chapter 1</toc-entry-content>
    <toc-entry-content-on-resumed>Continuation of Chapter 1</toc-entry-content-on-resumed>
    <toc-entry ref="1.1">
	  <toc-entry-content>Chapter 1.1</toc-entry-content>
	  <toc-entry ref="1.1.1">
        <toc-entry-content>Chapter 1.1.1</toc-entry-content>
      </toc-entry>
      <toc-entry ref="1.1.2">
        <toc-entry-content>Chapter 1.1.2</toc-entry-content>
      </toc-entry>
      <toc-entry ref="1.1.3">
        <toc-entry-content>Chapter 1.1.3</toc-entry-content>
      </toc-entry>
    </toc-entry>
    <toc-entry ref="1.2">
	  <toc-entry-content>Chapter 1.2</toc-entry-content>
	  <toc-entry ref="1.2.1">
        <toc-entry-content>Chapter 1.2.1</toc-entry-content>
      </toc-entry>
      <toc-entry ref="1.2.2">
        <toc-entry-content>Chapter 1.2.2</toc-entry-content>
      </toc-entry>
    </toc-entry>
  </toc-entry>
  <toc-entry ref="2">
    <toc-entry-content>Chapter 2</toc-entry-content>
    <toc-entry-content-on-resumed>Continuation of Chapter 2</toc-entry-content-on-resumed>
    <toc-entry ref="2.1">
	  <toc-entry-content>Chapter 2.1</toc-entry-content>
	  <toc-entry ref="2.1.1">
        <toc-entry-content>Chapter 2.1.1</toc-entry-content>
      </toc-entry>
      <toc-entry ref="2.1.2">
        <toc-entry-content>Chapter 2.1.2</toc-entry-content>
      </toc-entry>
    </toc-entry>
	
	<!-- here, Dotify applies a volume break in the content -->
	<!-- the second volume also has some resumed text of the previous chapter,
	     in other words, its content does not start with the block with id "2.2" -->
	
    <toc-entry ref="2.2">
	  <toc-entry-content>Chapter 2.2</toc-entry-content>
	  <toc-entry ref="2.2.1">
        <toc-entry-content>Chapter 2.1.1</toc-entry-content>
      </toc-entry>
      <toc-entry ref="2.2.2">
        <toc-entry-content>Chapter 2.2.2</toc-entry-content>
      </toc-entry>
    </toc-entry>
    <toc-entry ref="2.3">
	  <toc-entry-content>Chapter 2.3</toc-entry-content>
    </toc-entry>
  </toc-entry>
</toc-sequence>
```

The algorithm works as follows:
Take the previous `toc-entry`. In this case, 2.1.2
If it has a `toc-entry-content-on-resumed`, apply that.
Otherwise, take the parent `toc-entry` and see if it has a `toc-entry-on-resumed`. If so, apply that.
Should you arrive at the root (the `toc-sequence`), then there is no resumed text.

This is functionality which can't be described with just a `use-when` construction.
So I suggest to abandon the idea of a `use-when` and a `$resumed` variable, and use a `toc-entry-content-on-resumed`.

Reply:
```
<toc-sequence range="volume-incl-resumed">
  <toc-entry ref="1">
    <toc-entry-content use-when="not $child-resumed">Chapter 1</toc-entry-content>
    <toc-entry-content use-when="$child-resumed">Continuation of Chapter 1</toc-entry-content>
    <toc-entry ref="1.1">
      <toc-entry-content>Chapter 1.1</toc-entry-content>
      <toc-entry ref="1.1.1">
        <toc-entry-content>Chapter 1.1.1</toc-entry-content>
      </toc-entry>
      <toc-entry ref="1.1.2">
        <toc-entry-content>Chapter 1.1.2</toc-entry-content>
      </toc-entry>
      <toc-entry ref="1.1.3">
        <toc-entry-content>Chapter 1.1.3</toc-entry-content>
      </toc-entry>
    </toc-entry>
    <toc-entry ref="1.2">
      <toc-entry-content>Chapter 1.2</toc-entry-content>
      <toc-entry ref="1.2.1">
        <toc-entry-content>Chapter 1.2.1</toc-entry-content>
      </toc-entry>
      <toc-entry ref="1.2.2">
        <toc-entry-content>Chapter 1.2.2</toc-entry-content>
      </toc-entry>
    </toc-entry>
  </toc-entry>
  <toc-entry ref="2">
    <toc-entry-content use-when="not $child-resumed">Chapter 2</toc-entry-content>
    <toc-entry-content use-when="$child-resumed">Continuation of Chapter 2</toc-entry-content>
    <toc-entry ref="2.1">
      <toc-entry-content>Chapter 2.1</toc-entry-content>
      <toc-entry ref="2.1.1">
        <toc-entry-content>Chapter 2.1.1</toc-entry-content>
      </toc-entry>
      <toc-entry ref="2.1.2">
        <toc-entry-content>Chapter 2.1.2</toc-entry-content>
      </toc-entry>
    </toc-entry>
	
    <!-- here, Dotify applies a volume break in the content -->
    <!-- the second volume also has some resumed text of the previous chapter,
         in other words, its content does not start with the block with id "2.2" -->
	
    <toc-entry ref="2.2">
      <toc-entry-content use-when="not $resumed">Chapter 2.2</toc-entry-content>
      <toc-entry-content use-when="$resumed">Continuation of Chapter 2.2</toc-entry-content>
      <toc-entry ref="2.2.1">
        <toc-entry-content>Chapter 2.1.1</toc-entry-content>
      </toc-entry>
      <toc-entry ref="2.2.2">
        <toc-entry-content>Chapter 2.2.2</toc-entry-content>
      </toc-entry>
    </toc-entry>
    <toc-entry ref="2.3">
      <toc-entry-content>Chapter 2.3</toc-entry-content>
    </toc-entry>
  </toc-entry>
</toc-sequence>
```

Hi @PaulRambags

I just wanted to add an example with use-when syntax. Notice that $resumed could have the meaning $resumed = 'either this node is resumed or any of its children'. This would depreciate the $child-resumed above but I add it as a clarification.

Best regards
Daniel

Reply:
Hi @kalaspuffar,
I think you are using a condition (a `use-when`) to express specific behaviour. In your proposal, when `use-when` evaluates to `true`, you say that you should take the resumed text of any of its children. You are adding very specific functionality to the `use-when` clause. That can't be done.

Reply:
By the way, `toc-sequence` should be `table-of-contents` and it has no `range` attribute. I changed it in my example. I apologize for this.

Reply:
Hi @PaulRambags.

No, I don't say that you will take all its children. Just the ```use-when``` statements that evaluate to true.

Best regards
Daniel

Reply:
@kalaspuffar In this case you have a `use-when` which has a very specific behaviour. It does not what you would expect, namely, that it it a condition that expresses when to use the given `toc-entry-content`. It does more: when it evaluates to `true`, certain resumed functionality must be applied. In this way, you are using a general construction (the `use-when`) for a very specific purpose (the resumed functionality). I would strongly suggest not to add such specific functionality to the `use-when`.

Reply:
Hi @PaulRambags.

I think we have some missing information here. The only thing we have talked about is the general case where we need to use a text or not depending on an expression. If there is a case where you need more specialized behavior, then you need to describe this case for us to be a part of the discussion.

Best regards
Daniel

Reply:
Hi again.

As a side note, inheritance where never a part of our initial discussion, but in general, I don't see any problem with solving that case.

Best regards
Daniel

Reply:
I mentioned this concept, which @bertfrees named _inheritance_, already in the example in the original issue (see https://github.com/brailleapps/dotify.formatter.impl/issues/106#issuecomment-543045156). 
This concept of inheriance cannot be solved with merely a `use-when` and a `$resumed`, but it can be solved when you allow a `toc-entry-content-on-resumed`. With the `toc-entry-content-on-resumed` you don't need the `use-when` anymore and also you don't need the `$resumed`. So why further bother about the `use-when` and the `$resumed`?

Reply:
Hi @PaulRambags if I understand you correctly from the posts above, you are proposing to change what we agreed upon during the last call and also that it actually involves much more than just adding the possibility to have conditional content in the toc-entries. I believe we need to make a new call in that case to discuss this, since that would mean that we are basically back to square one. Would you like us to do this?

Reply:
Hi guys. I've been following the discussion from the sideline because as I said before both approaches are viable options to me and they both have some advantages and disadvantages so whatever is decided is fine for me.

I'm chiming in because I feel the discussion isn't really going anywhere.

Firstly, I think Paul is using the wrong argument to convince you to move to the other approach. It's not impossible to come up with a way to achieve this kind of behavior with the `use-when` approach, and of course Daniel has taken this up as a challenge :-) (I would do the same.) But moreover, I think it's not even so important that this behavior can be achieved. As I said before, it's also a solution to simply repeat `toc-entry-content`s in descendant elements whenever necessary. Nothing wrong with a bit of repetition.

I do understand what Daniel is trying to say with his example. However it makes things more complicated. You are selecting not only from the list of child `toc-entry-content` of the current `toc-entry`, but also from the `toc-entry-content` of all the ancestor `toc-entry`. I mean, it can be done like this, but then you loose most of the benefit of having a general solution IMO.

As far as I understand from Paul his problem with the `use-when` approach is more the principle that a very general solution is used for something that could benefit from some more specific behavior, while the generality of the solution does not offer much benefit because it's not very likely that the `use-when` will see other use cases.

I agree with Fredrik that a quick call to sort this out would be more effective.

Reply:
@bertfrees
>Firstly, I think Paul is using the wrong argument to convince you to move to the other approach. It's not impossible to come up with a way to achieve this kind of behavior with the use-when approach
I challenge you to define a default behaviour, that is, when there is no `use-when` clause. As you already mentioned, the default would be `not($resumed)` but then you are introducting a `use-when` which is by itself already resumed specific.

>Nothing wrong with a bit of repetition.
I would not call that _a bit_ of repetition. The amount of duplication is exponential. For instance, if there are 6 levels (as in the Pipeline) and each level has an average of 6 items, then you will have to duplicate the same text appox. 6^6 = 46656 times.

Reply:
@PaulRambags
The default behaviour when `use-when` is omited must be true, i.e. use always. Nothing else makes sense.

Reply:
@PaulRambags can you please show the exponential example?

Reply:
@fredrikschill 
>The default behaviour when use-when is omited must be true, i.e. use always. Nothing else makes sense.

In that case, when you introduce `use-when` as you suggested, you will have to add `use-when=(! $resumed)` everywhere to get the standard behavior. With standard behavior I mean how `toc-entry`s behave in the current implementation.

Reply:
> can you please show the exponential example?

Basically, you will have to add a resumed text at _every_ `toc-entry`.
Here is a part of the example above, and how it would be with a `use-when` and a `$resumed` and such that `use-when` evaluates to `true` by default:

```
<table-of-contents>
  ...
  <toc-entry ref="2">
    <toc-entry-content use-when="(! $resumed)">Chapter 2</toc-entry-content>
    <toc-entry-content use-when="$resumed">Continuation of Chapter 2</toc-entry-content-on-resumed>
    <toc-entry ref="2.1">
      <toc-entry-content use-when="(! $resumed)">Chapter 2.1</toc-entry-content>
      <toc-entry-content use-when="$resumed">Continuation of Chapter 2</toc-entry-content-on-resumed>
      <toc-entry ref="2.1.1">
        <toc-entry-content use-when="(! $resumed)">Chapter 2.1.1</toc-entry-content>
        <toc-entry-content use-when="$resumed">Continuation of Chapter 2</toc-entry-content-on-resumed>
      </toc-entry>
      <toc-entry ref="2.1.2">
        <toc-entry-content use-when="(! $resumed)">Chapter 2.1.2</toc-entry-content>
        <toc-entry-content use-when="$resumed">Continuation of Chapter 2</toc-entry-content-on-resumed>
      </toc-entry>
    </toc-entry>
	
    <!-- here, Dotify applies a volume break in the content -->
    <!-- the second volume also has some resumed text of the previous chapter,
         in other words, its content does not start with the block with id "2.2" -->
	
    <toc-entry ref="2.2">
      <toc-entry-content use-when="(! $resumed)">Chapter 2.2</toc-entry-content>
      <toc-entry-content use-when="$resumed">Continuation of Chapter 2</toc-entry-content-on-resumed>
      <toc-entry ref="2.2.1">
        <toc-entry-content use-when="(! $resumed)">Chapter 2.1.1</toc-entry-content>
        <toc-entry-content use-when="$resumed">Continuation of Chapter 2</toc-entry-content-on-resumed>
      </toc-entry>
      <toc-entry ref="2.2.2">
        <toc-entry-content use-when="(! $resumed)">Chapter 2.2.2</toc-entry-content>
        <toc-entry-content use-when="$resumed">Continuation of Chapter 2</toc-entry-content-on-resumed>
      </toc-entry>
    </toc-entry>
    <toc-entry ref="2.3">
      <toc-entry-content use-when="(! $resumed)">Chapter 2.3</toc-entry-content>
      <toc-entry-content use-when="$resumed">Continuation of Chapter 2</toc-entry-content-on-resumed>
    </toc-entry>
  </toc-entry>
</table-of-contents>
```

As you can see, the text "Continuation of Chapter 2" appears 8 times.
This small example has a TOC hierarchy with just 3 levels. If there would be more levels, the amount of times that you will have to insert this text in the TOC will increase exponentially.

Reply:
@PaulRambags Sure it is possible. Daniel's example is just one way to do it.

That the amount of repetition would be exponential is of course bollocks, you know it. You have to repeat a toc-entry-content at most once for every descendant toc-entry. How is that exponential? Exponential with what?

Anyway, let's end the bike-shedding now because it is getting us nowhere. I think Paul has made his point clear that the `use-when` approach is not all rosy. Now we need to make a decision and get on with our lives.

Reply:
> Exponential with what?

Suppose you want only a resumed text at the top level, as in my example. So only at levels 1, 2, 3, ... but not at 1.1, 1.2, etc.
If the depth of your TOC is `n` (in my example, `n` = 3 because there is no section 1.1.1.1 only a section 1.1.1) and at every level there are (in average) `m` items (in my example, `m` = 3 because there is a 2.1, 2.2 and 2.3, then you will have to repeat the same resumed text `m^(n - 1)` times, where `^` stands for "to the power of". So if the TOC depth increases, the number of repetitions increases exponentially.

Reply:
OK, interesting way of looking at it :-)
So a book with 6 heading levels has 46656 headings?